staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go

```diff
@@ -447,7 +447,19 @@ func (h *UpgradeAwareHandler) tryUpgrade(w http.ResponseWriter, req *http.Reques
 	// clean up the other half of the connection.
 	select {
 	case <-writerComplete:
+		// Once the client -> backend writer completes, we don't want to block
+		// forever, but also don't want to race in the case of network blips.
+		deadline := time.Now().Add(500 * time.Millisecond)
+		backendConn.SetReadDeadline(deadline)
+		requestHijackedConn.SetWriteDeadline(deadline)
+		<-readerComplete
```

Comment:
I think this can be simpler
```suggestion
		// give a chance to terminate gracefully or timeout
		// 0.5s is the default timeout used in socat that performs a similar operation
		// https://linux.die.net/man/1/socat
		select {
		   case  <-readerComplete:
		   case <-time.After(time.Duration(500) * time.Millisecond):
		        klog.Errorf("Connection didn't close gracefully proxying data from client to backend", err)
                }
```